lib/filewatch/watched_file.rb

```diff
@@ -0,0 +1,247 @@
+      [fileId, 0, 0] # dev_* doesn't make sense on Windows
+    end
+
+    def self.nix_inode(path, stat)
+      [stat.ino.to_s, stat.dev_major, stat.dev_minor]
+    end
+
+    def self.inode(path, stat)
+      # removed the send(sym, ...) call as
+      # its slow and this is called on each restat
+      if FILEWATCH_INODE_METHOD == :win_inode
+        win_inode(path, stat)
+      else
+        nix_inode(path, stat)
+      end
+    end
+
+    attr_reader :bytes_read, :state, :file, :buffer, :recent_states
+    attr_reader :path, :filestat, :accessed_at, :modified_at, :pathname
+    attr_reader :sdb_key_v1, :last_stat_size, :listener
+    attr_accessor :last_open_warning_at
+
+    # this class represents a file that has been discovered
+    def initialize(pathname, stat, settings)
+      @settings = settings
+      @pathname = Pathname.new(pathname) # given arg pathname might be a string or a Pathname object
+      @path = @pathname.to_path
+      @bytes_read = 0
+      @last_stat_size = 0
+      @sdb_key_v1 = InodeStruct.new(*self.class.inode(path, stat))
+      # initial as true means we have not associated this watched_file with a previous sincedb value yet.
+      # and we should read from the beginning if necessary
+      @initial = true
+      @recent_states = [] # keep last 8 states, managed in set_state
+      @state = :watched
+      set_stat(stat) # can change @last_stat_size
+      @listener = NullListener.new(@path)
+      @last_open_warning_at = nil
+      set_accessed_at
+    end
+
+    def set_listener(observer)
+      @listener = observer.listener_for(@path)
+    end
+
+    def unset_listener
+      @listener = nil
+    end
+
+    def has_listener?
+      !@listener.is_a?(NullListener)
+    end
+
+    def sincedb_key
+      @sdb_key_v1
+    end
+
+    def initial_completed
+      @initial = false
+    end
+
+    def set_accessed_at
+      @accessed_at = Time.now.to_f
+    end
+
+    def initial?
+      @initial
+    end
+
+    def compressed?
+      @path.end_with?('.gz','.gzip')
+    end
+
+    def size_changed?
+      @last_stat_size != bytes_read
+    end
+
+    def all_read?
+      @last_stat_size == bytes_read
+    end
+
+    def open
+      file_add_opened(FileOpener.open(@path))
+    end
+
+    def file_add_opened(rubyfile)
+      @file = rubyfile
+      @buffer = BufferedTokenizer.new(@settings.delimiter) if @buffer.nil?
+    end
+
+    def file_close
+      return if @file.nil? || @file.closed?
+      @file.close
+      @file = nil
+    end
+
+    def file_seek(amount, whence = IO::SEEK_SET)
+      @file.sysseek(amount, whence)
+    end
+
+    def file_read(amount)
+      set_accessed_at
+      @file.sysread(amount)
+    end
+
+    def file_open?
+      !@file.nil? && !@file.closed?
+    end
+
+    def reset_buffer
+      @buffer.input.clear
+    end
+
+    def buffer_extract(data)
+      @buffer.extract(data)
+    end
+
+    def buffer_delimiter_byte_size
+      @buffer.delimiter_byte_size
```

Comment:
AFAICT `BufferedTokenizer` does not have a `delimiter_byte_size` method. It looks like the original intent might have been `@buffer.delimiter.bytesize` ? hence the need to expose `attr_reader :delimiter` ? 
In any case, can't this be replaced with:
```
def buffer_delimiter_byte_size
  @settings.delimiter_byte_size
end
```
or
```
def buffer_delimiter_byte_size
  @settings.delimiter.bytesize
end
```
so that `attr_reader delimiter` is not required and we can reuse the Logstash `BufferedTokenizer` ?